src/main/java/com/mongodb/hibernate/internal/type/ObjectIdJdbcType.java

```diff
@@ -0,0 +1,125 @@
+    @Serial
+    private static final long serialVersionUID = 1L;
+
+    private static final MqlType MQL_TYPE = MqlType.OBJECT_ID;
+    private static final ObjectIdJavaType JAVA_TYPE = ObjectIdJavaType.INSTANCE;
+    public static final ObjectIdJdbcType INSTANCE = new ObjectIdJdbcType();
+
+    private ObjectIdJdbcType() {}
+
+    @Override
+    public int getJdbcTypeCode() {
+        return MQL_TYPE.getVendorTypeNumber();
+    }
+
+    @Override
+    public String getFriendlyName() {
+        return MQL_TYPE.getName();
+    }
+
+    @Override
+    public <X> ValueBinder<X> getBinder(JavaType<X> javaType) {
+        if (!javaType.equals(JAVA_TYPE)) {
+            throw new FeatureNotSupportedException();
+        }
+        @SuppressWarnings("unchecked")
+        var result = (ValueBinder<X>) Binder.INSTANCE;
+        return result;
+    }
+
+    @Override
+    public <X> ValueExtractor<X> getExtractor(JavaType<X> javaType) {
+        if (!javaType.equals(JAVA_TYPE)) {
+            throw new FeatureNotSupportedException();
+        }
+        @SuppressWarnings("unchecked")
+        var result = (ValueExtractor<X>) Extractor.INSTANCE;
+        return result;
+    }
+
+    private static final class Binder extends BasicBinder<ObjectId> {
+        @Serial
+        private static final long serialVersionUID = 1L;
+
+        static Binder INSTANCE = new Binder();
+
+        private Binder() {
+            super(JAVA_TYPE, ObjectIdJdbcType.INSTANCE);
+        }
+
+        @Override
+        protected void doBind(PreparedStatement st, ObjectId value, int index, WrapperOptions options)
+                throws SQLException {
+            st.setObject(index, value, getJdbcType().getJdbcTypeCode());
+        }
+
+        @Override
+        protected void doBind(CallableStatement st, ObjectId value, String name, WrapperOptions options)
+                throws SQLException {
+            throw new SQLFeatureNotSupportedException();
+        }
+    }
+
+    private static final class Extractor extends BasicExtractor<ObjectId> {
+        @Serial
+        private static final long serialVersionUID = 1L;
+
+        static Extractor INSTANCE = new Extractor();
+
+        private Extractor() {
+            super(JAVA_TYPE, ObjectIdJdbcType.INSTANCE);
+        }
+
+        @Override
+        protected ObjectId doExtract(ResultSet rs, int paramIndex, WrapperOptions options) throws SQLException {
+            return rs.getObject(paramIndex, getJavaType().getJavaTypeClass());
+        }
+
+        @Override
+        protected ObjectId doExtract(CallableStatement statement, int index, WrapperOptions options)
+                throws SQLException {
+            throw new SQLFeatureNotSupportedException();
+        }
+
+        @Override
+        protected ObjectId doExtract(CallableStatement statement, String name, WrapperOptions options)
+                throws SQLException {
+            throw new SQLFeatureNotSupportedException();
+        }
+    }
+}
```

Comment:
per Hibernate convention, no singleton Binder or Extractor is needed for `getBinder()` and `getExtractor()` would only be called once and only once.

Reply:
I failed to find that convention. Could you please point to the corresponding docs?

Reply:
no doc on this. It is simply some convention developers follow by habbit. Usually SessionFactory is the global context and if the instance is only created once, no need to bother with INSTANCE. But if multiple SessionFactories is in our scope, it is fine to use INSTANCE.